From b19a6b6bee9e571bc77818951cb4a4864e79a9d4 Mon Sep 17 00:00:00 2001
From: Akash Garg <gargakash@google.com>
Date: Thu, 26 Mar 2026 08:13:00 -0700
Subject: [PATCH] Prevent SDK Sandbox from bypassing isSystemApp check

The isSystemApp() check in CarrierConfigLoader incorrectly identifies
apps running within the SDK Sandbox as system apps, because their UIDs
resolve to the com.google.android.sdksandbox package (which is a system
package). This allows unprivileged apps in the SDK Sandbox to bypass
carrier restrictions on user builds by calling
overrideConfig(persistent=true).

This change adds an explicit Process.isSdkSandboxUid() check to ensure
SDK Sandbox UIDs are properly rejected.

Bug: 477022787
Test: Verified Manually and added result to bugreport
Flag: EXEMPT BUGFIX
Cherrypick-From:
https://googleplex-android-review.googlesource.com/q/commit:6f302b2c0b58732e8d5dd4587e78cefd3450226d
Merged-In: If7ce19d02a851a869586e3e03ed2e43e1edfd420
Change-Id: If7ce19d02a851a869586e3e03ed2e43e1edfd420
---
 .../android/phone/CarrierConfigLoader.java    | 31 ++++++++++++++-----
 .../phone/CarrierConfigLoaderTest.java        | 18 ++++++++++-
 2 files changed, 41 insertions(+), 8 deletions(-)

diff --git a/src/com/android/phone/CarrierConfigLoader.java b/src/com/android/phone/CarrierConfigLoader.java
index 0963a6543..21e8ffa84 100644
--- a/src/com/android/phone/CarrierConfigLoader.java
+++ b/src/com/android/phone/CarrierConfigLoader.java
@@ -46,6 +46,7 @@ import android.os.Looper;
 import android.os.Message;
 import android.os.PermissionEnforcer;
 import android.os.PersistableBundle;
+import android.os.Process;
 import android.os.RemoteException;
 import android.os.ResultReceiver;
 import android.os.SystemProperties;
@@ -1472,7 +1473,7 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
 
         // Do not allow shell UID to override the carrier config. This will not impact
         // the CTS and telephony shell commands as they use different uids
-        if (TelephonyPermissions.isShell(getCallingUid())) {
+        if (TelephonyPermissions.isShell(getBinderCallingUid())) {
             throw new SecurityException("overrideConfig cannot be invoked by shell");
         }
 
@@ -1520,8 +1521,14 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
 
     private boolean isSystemApp() {
         try {
+            int callingUid = getBinderCallingUid();
+            if (isSdkSandboxUidInternal(callingUid)) {
+                loge("isSystemApp: rejected call from SDK sandbox with uid=" + callingUid);
+                return false;
+            }
+
             String callingPackage = mContext.getPackageManager().getNameForUid(
-                    Binder.getCallingUid());
+                    callingUid);
 
             ApplicationInfo appInfo = mContext.getPackageManager().getApplicationInfo(
                     callingPackage, 0);
@@ -1575,7 +1582,7 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
         // from clearing the cache is passed back to the carrier app. With the files successfully
         // deleted, this can return and we will eventually bind to the carrier app.
         String callingPackageName = mContext.getPackageManager().getNameForUid(
-                Binder.getCallingUid());
+                getBinderCallingUid());
         clearCachedConfigForPackage(callingPackageName);
         mNeedNotifyCallback[phoneId] = true;
         updateConfigForPhoneId(phoneId);
@@ -1715,8 +1722,8 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
         IndentingPrintWriter indentPW = new IndentingPrintWriter(pw, "    ");
         if (mContext.checkCallingOrSelfPermission(android.Manifest.permission.DUMP)
                 != PackageManager.PERMISSION_GRANTED) {
-            indentPW.println("Permission Denial: can't dump carrierconfig from from pid="
-                    + Binder.getCallingPid() + ", uid=" + Binder.getCallingUid());
+            indentPW.println("Permission Denial: can't dump carrierconfig from pid="
+                    + Binder.getCallingPid() + ", uid=" + getBinderCallingUid());
             return;
         }
         String requestingPackage = null;
@@ -1822,7 +1829,7 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
      */
     private void enforceCallerIsSystemOrRequestingPackage(@NonNull String requestingPackage)
             throws SecurityException {
-        final int callingUid = Binder.getCallingUid();
+        final int callingUid = getBinderCallingUid();
         if (TelephonyPermissions.isRootOrShell(callingUid)
                 || TelephonyPermissions.isSystemOrPhone(
                 callingUid)) {
@@ -1921,6 +1928,16 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
         indentPW.decreaseIndent();
     }
 
+    @VisibleForTesting
+    protected int getBinderCallingUid() {
+        return Binder.getCallingUid();
+    }
+
+    @VisibleForTesting
+    protected boolean isSdkSandboxUidInternal(int uid) {
+        return Process.isSdkSandboxUid(uid);
+    }
+
     private boolean hasCarrierPrivileges(@NonNull String pkgName, int phoneId) {
         int subId = SubscriptionManager.getSubscriptionId(phoneId);
         if (!SubscriptionManager.isValidSubscriptionId(subId)) {
@@ -1941,7 +1958,7 @@ public class CarrierConfigLoader extends ICarrierConfigLoader.Stub {
             PackageManager pm = mContext.createContextAsUser(Binder.getCallingUserHandle(), 0)
                     .getPackageManager();
             if (pm == null) return null;
-            String[] callingPackageNames = pm.getPackagesForUid(Binder.getCallingUid());
+            String[] callingPackageNames = pm.getPackagesForUid(getBinderCallingUid());
             return (callingPackageNames == null) ? null : callingPackageNames[0];
         }
         if (mPackageManager == null) return null;
diff --git a/tests/src/com/android/phone/CarrierConfigLoaderTest.java b/tests/src/com/android/phone/CarrierConfigLoaderTest.java
index 5b306e6a6..4cef460ad 100644
--- a/tests/src/com/android/phone/CarrierConfigLoaderTest.java
+++ b/tests/src/com/android/phone/CarrierConfigLoaderTest.java
@@ -107,6 +107,11 @@ public class CarrierConfigLoaderTest extends TelephonyTestBase {
     private TelephonyManager mTelephonyManager;
     private CarrierConfigLoader mCarrierConfigLoader;
     private Handler mHandler;
+    private int mCapturedCarrierId;
+    private int mCapturedCarrierServiceUid;
+    private int[] mCapturedPackageUids;
+    private int mFakeCallingUid;
+    private boolean mFakeIsSdkSandboxUid;
 
     // The AIDL stub will use PermissionEnforcer to check permission from the caller.
     private FakePermissionEnforcer mFakePermissionEnforcer = new FakePermissionEnforcer();
@@ -151,7 +156,18 @@ public class CarrierConfigLoaderTest extends TelephonyTestBase {
                 mTelephonyRegistryManager);
 
         mCarrierConfigLoader = new CarrierConfigLoader(mContext, mTestLooper,
-                mFeatureFlags);
+                mFeatureFlags) {
+
+                @Override
+                protected int getBinderCallingUid() {
+                        return mFakeCallingUid;
+                }
+
+                @Override
+                protected boolean isSdkSandboxUidInternal(int uid) {
+                        return mFakeIsSdkSandboxUid;
+                }
+        };
         mHandler = mCarrierConfigLoader.getHandler();
 
         // Clear all configs to have the same starting point.
-- 
2.34.1

